### DISCUSSION TITLE
`record` and `record?` generic constraints

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#7528] [Language Ideas] [answered]
A couple of times I've written generic methods and wanted access to the `with` keyword, in order to copy `T` so as not to mutate the original.

As a workaround I had to reflect into `object.MemberwiseClone()`, which was enough for my use case but it's not an optimal solution.

usage:
```csharp

public static T RemoveId<T>(T obj) where T : record, IHasId
{
    return obj with { Id = null };
}
```

There's been discussion of expanding the `with` keyword beyond records, but this is a much less invasive and much less opinionated feature.

**Comment by HaloFour** [comment]
Without redesigning how `with` works this wouldn't be possible to manage via generic constraints because there is no common method used to clone the record.  Each record defines its own independently.

**Reply by RenderMichael** [reply]
much like the `new()` constraint uses `Activator.CreateInstance()` under the hood, maybe this can just reflect into the `<Clone>$` method

**Reply by HaloFour** [reply]
That's an implementation detail that may change in the future.

**Reply by RenderMichael** [reply]
`record`s defined against one SDK does not need to work on a different SDK right? Since it doesn't use a source-generator or anything.

So if `<Clone>$` is an implementation detail that changes, the `record` generic parameter implementation can change as well. Though I can see how this may limit the wiggle room available.

Especially given the `UnsafeAccessorAttribute`  shipped in .NET 8, reflection of a well-known member can be performant.

**Comment by HaloFour** [comment, marked as answer]
Also, see: https://github.com/dotnet/csharplang/discussions/3859

IMO, the teams effort would be better spent describing "withers" as a pattern that can be supported by generic type constraints, rather than adding one specific generic type constraint specific to the current implementation of records.

**Reply by RenderMichael** [reply]
That's fair enough. It would be nice if the "wither" pattern could fit a generic constraint as well

**Comment by TonyValenti** [comment]
I have an empty base class called DisplayRecord that all my records inherit from.  I can then use constraints and get WITH.
